torch/lib/c10d/TCPStore.cpp

```diff
@@ -57,11 +56,173 @@ TCPStoreDaemon::~TCPStoreDaemon() {
   closeStopSignal();
 }
 
-void TCPStoreDaemon::join() {
+void BackgroundThread::join() {
   daemonThread_.join();
 }
 
-void TCPStoreDaemon::queryFds(std::vector<struct pollfd>& fds) {
+#ifdef _WIN32
+void BackgroundThread::initStopSignal() {
+  ghStopEvent_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (ghStopEvent_ == NULL) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  CloseHandle(ghStopEvent_);
+}
+
+void BackgroundThread::stop() {
+  SetEvent(ghStopEvent_);
+}
+#else
+void BackgroundThread::initStopSignal() {
+  if (pipe(controlPipeFd_.data()) == -1) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "BackgroundThread run");
+  }
+}
+
+void BackgroundThread::closeStopSignal() {
+  for (auto fd : controlPipeFd_) {
+    if (fd != -1) {
+      ::close(fd);
+    }
+  }
+}
+
+void BackgroundThread::stop() {
+  if (controlPipeFd_[1] != -1) {
+    // close the write end of the pipe
+    ::close(controlPipeFd_[1]);
+    controlPipeFd_[1] = -1;
+  }
+}
+#endif
+
+// TCPStoreListener class methods
+TCPStoreWorkerDaemon::TCPStoreWorkerDaemon(int listenSocket)
+    : BackgroundThread(listenSocket) {
+  daemonThread_ = std::thread(&TCPStoreWorkerDaemon::run, this);
+}
+
+void TCPStoreWorkerDaemon::addCallback(
+    std::string key,
+    WatchKeyCallback callback) {
+  const std::lock_guard<std::mutex> lock(_keyToCallbacksMutex);
+  _keyToCallbacks[key] = callback;
+}
+
+// Runs all the callbacks that the worker has registered
+void TCPStoreWorkerDaemon::callbackHandler(int socket) {
+  auto watchResponse = tcputil::recvValue<WatchResponseType>(socket);
+  if (watchResponse == WatchResponseType::KEY_CALLBACK_REGISTERED) {
+    // Notify the waiting "watchKey" operation to return
+    setCallbackRegisteredData(true);
```

Comment:
this can run concurrently with another invocation in `TCPStore::watchKey`. Will this be a problem? Say it is set to true here, and then immediately flipped to false by `TCPStore::watchKey`. And then the next line `getCallbackRegisteredCondVar().notify_one();` kicks in?

Reply:
addressed offline. This won't occur, as TCPStore::watchKey will block until this is done before processing the confirmation from the master.